File: Backend/ClearChain.Domain/Entities/Organization.cs
namespace ClearChain.Domain.Entities;

public class Organization
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Type { get; set; } = string.Empty; // grocery, ngo, admin
    public string Email { get; set; } = string.Empty;
    public string? PasswordHash { get; set; }

    public string? Phone { get; set; }
    public string? Address { get; set; }
    public string? Location { get; set; }
    public string? State { get; set; }
    public string? ZipCode { get; set; }

    public bool Verified { get; set; }
    public string VerificationStatus { get; set; } = "pending";
    public string? VerificationNotes { get; set; }   // Admin's note when approving/rejecting (e.g. rejection reason)
    public string? Hours { get; set; }

    public string AuthProvider { get; set; } = "local";
    public string? ProfilePictureUrl { get; set; }

    public double? Latitude { get; set; }
    public double? Longitude { get; set; }
    public string? ContactPerson { get; set; }       // NGO only
    public string? PickupInstructions { get; set; }   // Grocery only
    public string? Description { get; set; }          // Both roles

    // ── Auth security ────────────────────────────────────────────────────────
    public int FailedLoginCount { get; set; } = 0;
    public DateTime? LockoutUntil { get; set; }
    public bool IsDeleted { get; set; } = false;
    public DateTime? DeletedAt { get; set; }

    // ── Email verification ───────────────────────────────────────────────────
    public bool EmailVerified { get; set; } = false;
    public string? EmailVerificationToken { get; set; }
    public DateTime? EmailVerificationTokenExpiry { get; set; }

    // ── Onboarding document ──────────────────────────────────────────────────
    public string? DocumentUrl { get; set; }        // Verification doc (business reg / charity cert)
    public string? DocumentMimeType { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

File: Mobile/app/src/main/java/com/clearchain/app/domain/model/Organization.kt
package com.clearchain.app.domain.model

import android.annotation.SuppressLint
import kotlinx.serialization.Serializable

@SuppressLint("UnsafeOptInUsageError")
@Serializable
data class Organization(
    val id: String,
    val name: String,
    val type: OrganizationType,
    val email: String,
    val phone: String,
    val address: String,
    val location: String,
    val verified: Boolean,
    val verificationStatus: VerificationStatus,
    val verificationNotes: String? = null,
    val hours: String? = null,
    val profilePictureUrl: String? = null,
    val createdAt: String,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val state: String? = null,
    val zipCode: String? = null,
    val contactPerson: String? = null,
    val pickupInstructions: String? = null,
    val description: String? = null,
    // ── Document fields ──────────────────────────────────────────────────────
    val documentUrl: String? = null,
    val documentMimeType: String? = null
) {
    fun isProfileComplete(): Boolean {
        val commonComplete = phone.isNotBlank() &&
            address.isNotBlank() &&
            location.isNotBlank() &&
            !hours.isNullOrBlank()

        return when (type) {
            OrganizationType.NGO -> commonComplete && !contactPerson.isNullOrBlank()
            OrganizationType.GROCERY -> commonComplete && !contactPerson.isNullOrBlank()
            OrganizationType.ADMIN -> true
        }
    }

    /**
     * True when the org has finished onboarding but an admin has not approved it yet
     * (pending or rejected). Such orgs are held on the PendingReview screen and blocked
     * from creating listings / pickup requests. Admins are never gated.
     */
    fun requiresVerificationGate(): Boolean =
        type != OrganizationType.ADMIN && verificationStatus != VerificationStatus.APPROVED

}

@Serializable
enum class OrganizationType { GROCERY, NGO, ADMIN }

@Serializable
enum class VerificationStatus { PENDING, APPROVED, REJECTED }
